### ISSUE TITLE
calculate matrix builds dynamically based on source code inspection

### ISSUE THREAD

**Issue by macb** [drone/drone#1362]
Matrix builds was one of the most exciting Drone 0.4 features when first evaluating the switch. At DigitalOcean we have a large single repo for our chef codebase which includes quite a few cookbooks. It looks something like:

```
.drone.yml
script/
  cibuild
  bootstrap
cookbooks/
  foo/
    spec/
  bar/
    spec/
  baz/
    spec/
```

We have a single drone file at the root which runs the bootstrap and cibuild scripts. The cibuild script determines what cookbooks (if any) have changed and then proceeds to to run the specs for those cookbooks. This results in long-ish build times as each suite processes serially and the output becomes very hard to parse as the number of suites increases (sometimes surpassing the UI truncation limit).

If instead we apply the idea of matrix builds, we could have each cookbook suite process as its own build with a different directory to work inside of.

This would mean as part of receiving a push/pr hook, we would need to be able to interact with the changeset. It would appear this would require access to the repository changed as it is not currently possible via the API (at least for github).

It gets further complicated if you think about wanting to then apply a matrix build to each of those changes. Imagine a scenario where we changed files in `cookbooks/foo` and `cookbooks/bar`. We'll want to run 2 builds, one for `foo` and one for `bar`. But we also specify a `matrix` field in our drone yaml:

```
matrix:
  RUBY_VERSION:
    - 2.1
    - 2.2
```

So now we want to run 4 different builds:

```
foo @ 2.1
foo @ 2.2
bar @ 2.1
bar @ 2.2
```

**Comment by bradrydzewski** [comment]
Unfortunately this is not possible given the current drone architecture and would be difficult to implement at this point. I am therefore closing this as out of scope, but appreciate the feedback.
